### ISSUE TITLE
Need support for AMD Ryzen Embedded V1000/APU

### ISSUE THREAD

**Issue by jarvis-hal** [RadeonOpenCompute/hcc#883]
Currently,  AMD Ryzen Embedded V1000 and APU are not supported by HCC which does not allow Tensorflow to take advantage of the APU power coming from AMD Ryzen Embedded V1000 series. This could be hindering many low power AI applications which V1000 could enable. Please help to add support.

**Comment by dignakov** [comment]
Would it be possible to get some feedback on whether this is something being worked on/considered?

**Comment by dignakov** [comment]
Just wanted to bump this one more time to see if this is being considered?

**Comment by whchung** [comment]
I have one V1000 on my cube, and I do want to see embedded APU be put onto ROCm roadmap but unfortunately there is no ETA for this. Notice this is not really an HCC issue but a Linux driver-level issue so the best place is https://github.com/RadeonOpenCompute/ROCm/issues/435.

**Comment by dignakov** [comment]
Thanks, I was looking at that thread too. I ended up here because of a comment by kentrussell:

> That's something you should bring up in the HCC github Bug Reports (https://github.com/RadeonOpenCompute/hcc). ROCK/ROCT/ROCR support APUs, but HCC doesn't. If you want to get TensorFlow support for APUs, I'd bring it to their attention directly. The base has support (kernel/thunk/runtime), but HCC obviously needs to support it to get in there

And the person who made this thread was participating in that one too.

**Comment by whchung** [comment]
@EvilPictureBook thanks. I didn't know kernel side of things may have been cleared up. Not so sure if CRAT table on BIOS for V1000 though.

Say in case one's able to run `/opt/rocm/bin/rocminfo` on V1000, which validates if ROCK/ROCT/ROCR truly works on the system, then enabling codegen for it is relatively an easy task in HCC:

Clang driver:
https://github.com/RadeonOpenCompute/hcc-clang-upgrade/blob/clang_tot_upgrade/lib/Driver/ToolChains/Hcc.cpp#L175

Link with ROCm-Device-Libs and invoking LLVM backend:
https://github.com/RadeonOpenCompute/hcc/blob/clang_tot_upgrade/lib/clamp-device.in#L151

Need to change those 2 places to ensure proper GCN ISA version are passed to match ISA version for V1000, which could be retrieved thru `/opt/rocm/bin/rocminfo`.

Unfortunately I haven't been able to run `/opt/rocm/bin/rocminfo` at least on my system yet, which led me to wonder it be a lower-level (ROCK/ROCT/ROCR / BIOS) issue.

**Comment by freddybc** [comment]
It is possible to get rocm working on APUs.
You can read about the effort here, https://bruhnspace.com/en/bruhnspace-rocm-for-amd-apus/
and download deb files from https://bruhnspace.com/rocm-apu/.

ROCm 2.6 built with APU support (gfx801, carrizo, bristol ridge, and gfx902 Raven Ridge).
rocm 2.7 is a mess... will build 2.8/2.9 if the ongoing renaming chaos has settled down.

**Comment by jt-lu** [comment]
> It is possible to get rocm working on APUs.
> You can read about the effort here, https://bruhnspace.com/en/bruhnspace-rocm-for-amd-apus/
> and download deb files from https://bruhnspace.com/rocm-apu/.
> 
> ROCm 2.6 built with APU support (gfx801, carrizo, bristol ridge, and gfx902 Raven Ridge).
> rocm 2.7 is a mess... will build 2.8/2.9 if the ongoing renaming chaos has settled down.

I get the following error when running TensorFlow benchmark on gfx902, ROCm 2.6

2019-09-27 17:39:13.808432: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1651] Found device 0 with properties: 
name: AMD Ryzen Embedded R1606G with Radeon Vega Gfx
AMDGPU ISA: gfx902
memoryClockRate (GHz) 1.2
pciBusID 0000:03:00.0
2019-09-27 17:39:13.808512: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library librocblas.so
2019-09-27 17:39:13.808526: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libMIOpen.so
2019-09-27 17:39:13.808538: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library librocfft.so
2019-09-27 17:39:13.808549: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library librocrand.so
2019-09-27 17:39:13.808612: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2019-09-27 17:39:13.808629: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-09-27 17:39:13.808636: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0 
2019-09-27 17:39:13.808642: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N 
2019-09-27 17:39:13.808730: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 11468 MB memory) -> physical GPU (device: 0, name: AMD Ryzen Embedded R1606G with Radeon Vega Gfx, pci bus id: 0000:03:00.0)
2019-09-27 17:39:14.678663: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line flag, not via TF_XLA_FLAGS) or set the envvar XLA_FLAGS=--xla_hlo_profile.
INFO:tensorflow:Running local_init_op.
I0927 17:39:43.430357 140018969397056 session_manager.py:500] Running local_init_op.
INFO:tensorflow:Done running local_init_op.
I0927 17:39:46.584101 140018969397056 session_manager.py:502] Done running local_init_op.
Running warm up
2019-09-27 17:39:48.691965: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library librocblas.so
2019-09-27 17:39:48.701037: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libMIOpen.so
MIOpen(HIP): miopenStatus_t miopenCreateTensorDescriptor(miopenTensorDescriptor_t *){
MIOpen(HIP): 	tensorDesc = 0
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenSetTensorDescriptor(miopenTensorDescriptor_t, miopenDataType_t, int, int *, int *){
MIOpen(HIP): 	tensorDesc = 
MIOpen(HIP): 	dataType = 1
MIOpen(HIP): 	nbDims = 4
MIOpen(HIP): 	dimsA = 32
MIOpen(HIP): 	stridesA = 158700
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenCreateTensorDescriptor(miopenTensorDescriptor_t *){
MIOpen(HIP): 	tensorDesc = 0
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenSetTensorDescriptor(miopenTensorDescriptor_t, miopenDataType_t, int, int *, int *){
MIOpen(HIP): 	tensorDesc = 
MIOpen(HIP): 	dataType = 1
MIOpen(HIP): 	nbDims = 4
MIOpen(HIP): 	dimsA = 64
MIOpen(HIP): 	stridesA = nullptr
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenCreateTensorDescriptor(miopenTensorDescriptor_t *){
MIOpen(HIP): 	tensorDesc = 0
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenSetTensorDescriptor(miopenTensorDescriptor_t, miopenDataType_t, int, int *, int *){
MIOpen(HIP): 	tensorDesc = 
MIOpen(HIP): 	dataType = 1
MIOpen(HIP): 	nbDims = 4
MIOpen(HIP): 	dimsA = 32
MIOpen(HIP): 	stridesA = 802816
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenCreateConvolutionDescriptor(miopenConvolutionDescriptor_t *){
MIOpen(HIP): 	convDesc = 0
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenInitConvolutionNdDescriptor(miopenConvolutionDescriptor_t, int, int *, int *, int *, miopenConvolutionMode_t){
MIOpen(HIP): 	convDesc = conv2d, miopenConvolution, miopenPaddingDefault, {0, 0}, {1, 1}, {1, 1}, 
MIOpen(HIP): 	spatialDim = 2
MIOpen(HIP): 	padA = 0
MIOpen(HIP): 	strideA = 2
MIOpen(HIP): 	dilationA = 1
MIOpen(HIP): 	c_mode = 0
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenSetConvolutionGroupCount(miopenConvolutionDescriptor_t, int){
MIOpen(HIP): 	convDesc = conv2d, miopenConvolution, miopenPaddingDefault, {0, 0}, {2, 2}, {1, 1}, 
MIOpen(HIP): 	groupCount = 1
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenConvolutionForwardGetWorkSpaceSize(miopenHandle_t, const miopenTensorDescriptor_t, const miopenTensorDescriptor_t, const miopenConvolutionDescriptor_t, const miopenTensorDescriptor_t, size_t *){
MIOpen(HIP): 	handle = stream: 0x56281a3c4200, device_id: 0
MIOpen(HIP): 	wDesc = 64, 3, 7, 7
MIOpen(HIP): 	yDesc = 32, 64, 112, 112
MIOpen(HIP): 	convDesc = conv2d, miopenConvolution, miopenPaddingDefault, {0, 0}, {2, 2}, {1, 1}, 
MIOpen(HIP): 	workSpaceSize = 140011613689616
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopenFindConvolutionForwardAlgorithm(miopenHandle_t, const miopenTensorDescriptor_t, const void *, const miopenTensorDescriptor_t, const void *, const miopenConvolutionDescriptor_t, const miopenTensorDescriptor_t, void *, const int, int *, miopenConvAlgoPerf_t *, void *, size_t, bool){
MIOpen(HIP): 	handle = stream: 0x56281a3c4200, device_id: 0
MIOpen(HIP): 	xDesc = 32, 3, 230, 230
MIOpen(HIP): 	x = 0x7f541f379d00
MIOpen(HIP): 	wDesc = 64, 3, 7, 7
MIOpen(HIP): 	w = 0x7f5419f26c00
MIOpen(HIP): 	convDesc = conv2d, miopenConvolution, miopenPaddingDefault, {0, 0}, {2, 2}, {1, 1}, 
MIOpen(HIP): 	yDesc = 32, 64, 112, 112
MIOpen(HIP): 	y = 0x7f54206d9300
MIOpen(HIP): 	requestAlgoCount = 1
MIOpen(HIP): 	returnedAlgoCount = 32598
MIOpen(HIP): 	perfResults = 
MIOpen(HIP): 	workSpace = 0x7f5419f2ff00
MIOpen(HIP): 	workSpaceSize = 7375872
MIOpen(HIP): 	exhaustiveSearch = 0
MIOpen(HIP): }
MIOpen(HIP): void miopen::DirConvFindCore(miopen::Handle &, const miopen::TensorDescriptor &, ConstData_t, const miopen::TensorDescriptor &, ConstData_t, const miopen::TensorDescriptor &, Data_t, Data_t, size_t, const miopen::ConvolutionDescriptor &, bool, miopen::DbRecord &){
MIOpen(HIP): 	"convolution, non 1x1" = convolution, non 1x1
MIOpen(HIP): }
MIOpen Error: /tmp/tmp.8FR64PNbKc/MIOpen/src/tmp_dir.cpp:18: Can't execute cd /tmp/miopen-MIOpenIm2d2Col.cl-a219-414e-fad2-6e85; /opt/rocm/bin/clang-ocl  -DNUM_CH_PER_WG=1 -DNUM_IM_BLKS_X=4 -DNUM_IM_BLKS=56 -DLOCAL_MEM_SIZE=1449 -DSTRIDE_GT_1=1 -DTILE_SZ_X=32 -DTILE_SZ_Y=8 -DUSE_IM_OFF_GUARD=1 -DMIOPEN_USE_FP16=0 -DMIOPEN_USE_FP32=1 -DMIOPEN_USE_BFP16=0  -DMIOPEN_USE_INT8=0 -DMIOPEN_USE_INT8x4=0 -mcpu=gfx902 -Wno-everything MIOpenIm2d2Col.cl -o /tmp/miopen-MIOpenIm2d2Col.cl-a219-414e-fad2-6e85/MIOpenIm2d2Col.cl.o
2019-09-27 17:39:48.737135: F tensorflow/stream_executor/rocm/rocm_dnn.cc:2707] Check failed: status == miopenStatusSuccess (7 vs. 0)Unable to find a suitable algorithm for doing forward convolution
Fatal Python error: Aborted

**Comment by AGenchev** [comment]
I tried their (https://bruhnspace.com) .deb files on docker image of Ubuntu 18 (Under Debian "Bullseye"  with kfd working). The hardware was Ryzen 5 PRO 3350G. Got it to the point where TF and PyTorch were reporting CUDA support on the APU. But any execution even of the simplest op on the GPU **failed**. For this reason, I'm waiting the prices of old Tesla M40 to drop a bit.
